woocommerce/woocommerce
# blueprint - add rest api

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR adds blueprint REST API endpoints.

POST /wp-json/blueprint/export
POST /wp-json/blueprint/import



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new JN site with this branch. Make sure to install `WooCommerce Beta Tester`
2. Skip the core profiler.
3. Navigate to `Tools -> WC Test Helper -> Features` and enable `blueprint` feature.

To test the endpoints using Curl, we'll need to modify the permission and nonce check.

1. Navigate to https://YOUR-JN-SITE/wp-admin/plugin-editor.php
2. Select `WooCommerce`
3. Find `src/Admin/Features/Blueprint/RestApi.php`
4. Change `permission_callback` to `__return_true` for both endpoints.
5. Scroll down to `import` method.
6. Remove nonce check `if` statement.

Testing with Curl

1. Open terminal in your local.
2. Run the following command. It should save test.json in your local. Please replace `YOUR-JN-SITE` with your JN site.

```
curl -X POST 'https://YOUR-JN-SITE/wp-json/blueprint/export' | sed -n 's/.*"data":\({.*}\),"type".*/\1/p' > test.json
```
3. Confirm the JSON is valid.
4. Run the following JSON. Place `ABSOLUTE-PATH-TO-JSON-FILE` and -YOUR-JN-SITE'

```
curl -X POST -F 'file=@ABSOLUTE-PATH-TO-JSON-FILE'  'https://YOUR-JN-SITE/wp-json/blueprint/export'
```

Both Curl requests should work without any errors.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>



#### Comment 
This is a sub-PR for add/blueprint-package-1

</details>
